File: frontend/src/services/projectService.ts
import { AuthService } from './authService'

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:3001'

export interface Project {
  id: number
  name: string
  description: string
  questionsAnswered: number
  totalQuestions: number
  completionRatio: number
  lastStudied: string | null
  createdAt: string
  isPublic: boolean
}

export interface CreateProjectRequest {
  name: string
  description?: string
  isPublic?: boolean
}

export interface UpdateProjectRequest {
  name?: string
  description?: string
  isPublic?: boolean
}

/**
 * Project service for managing user projects
 * Handles CRUD operations and project statistics
 */
export class ProjectService {
  /**
   * Get all user projects
   */
  static async getUserProjects(): Promise<Project[]> {
    const response = await fetch(`${API_BASE_URL}/api/projects`, {
      headers: {
        ...AuthService.getAuthHeader(),
      },
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error || 'Failed to fetch projects')
    }

    return data.projects
  }

  /**
   * Get a single project by ID
   */
  static async getProject(projectId: number): Promise<Project> {
    const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}`, {
      headers: {
        ...AuthService.getAuthHeader(),
      },
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error || 'Failed to fetch project')
    }

    return data.project
  }

  /**
   * Create a new project
   */
  static async createProject(projectData: CreateProjectRequest): Promise<Project> {
    const response = await fetch(`${API_BASE_URL}/api/projects`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...AuthService.getAuthHeader(),
      },
      body: JSON.stringify(projectData),
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error || 'Failed to create project')
    }

    return data.project
  }

  /**
   * Update a project
   */
  static async updateProject(projectId: number, updateData: UpdateProjectRequest): Promise<Project> {
    const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        ...AuthService.getAuthHeader(),
      },
      body: JSON.stringify(updateData),
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error || 'Failed to update project')
    }

    return data.project
  }

  /**
   * Delete a project
   */
  static async deleteProject(projectId: number): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}`, {
      method: 'DELETE',
      headers: {
        ...AuthService.getAuthHeader(),
      },
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error || 'Failed to delete project')
    }
  }

  /**
   * Get public projects for discovery
   */
  static async getPublicProjects(limit: number = 20): Promise<Project[]> {
    const response = await fetch(`${API_BASE_URL}/api/projects/public?limit=${limit}`)

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error || 'Failed to fetch public projects')
    }

    return data.projects
  }
}

File: backend-springboot/src/main/java/com/learningadvisor/backend/service/ProjectService.java
package com.learningadvisor.backend.service;

import com.learningadvisor.backend.dto.CreateProjectRequest;
import com.learningadvisor.backend.dto.ProjectDTO;
import com.learningadvisor.backend.dto.UpdateProjectRequest;
import com.learningadvisor.backend.entity.Project;
import com.learningadvisor.backend.entity.SessionAnswer;
import com.learningadvisor.backend.entity.StudySession;
import com.learningadvisor.backend.repository.ProjectRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Service layer for project management operations
 * Handles CRUD operations and statistics calculation
 */
@Service
@RequiredArgsConstructor
public class ProjectService {
    private final ProjectRepository projectRepository;

    /**
     * Create a new project for a user
     */
    @Transactional
    public ProjectDTO createProject(Long userId, CreateProjectRequest request) {
        Project project = new Project();
        project.setName(request.getName());
        project.setDescription(request.getDescription());
        project.setIsPublic(request.getIsPublic() != null ? request.getIsPublic() : false);
        project.setOwnerId(userId);

        Project savedProject = projectRepository.save(project);
        return convertToDTO(savedProject);
    }

    /**
     * Get all projects for a user with statistics
     */
    @Transactional(readOnly = true)
    public List<ProjectDTO> getUserProjects(Long userId) {
        List<Project> projects = projectRepository.findByOwnerIdOrderByCreatedAtDesc(userId);
        return projects.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Get a single project by ID with ownership check
     */
    @Transactional(readOnly = true)
    public Optional<ProjectDTO> getProjectById(Long projectId, Long userId) {
        return projectRepository.findByIdAndOwnerId(projectId, userId)
                .map(this::convertToDTO);
    }

    /**
     * Update a project
     */
    @Transactional
    public Optional<ProjectDTO> updateProject(Long projectId, Long userId, UpdateProjectRequest request) {
        Optional<Project> projectOpt = projectRepository.findByIdAndOwnerId(projectId, userId);

        if (projectOpt.isEmpty()) {
            return Optional.empty();
        }

        Project project = projectOpt.get();

        if (request.getName() != null) {
            project.setName(request.getName());
        }
        if (request.getDescription() != null) {
            project.setDescription(request.getDescription());
        }
        if (request.getIsPublic() != null) {
            project.setIsPublic(request.getIsPublic());
        }

        Project updatedProject = projectRepository.save(project);
        return Optional.of(convertToDTO(updatedProject));
    }

    /**
     * Delete a project
     */
    @Transactional
    public boolean deleteProject(Long projectId, Long userId) {
        Optional<Project> projectOpt = projectRepository.findByIdAndOwnerId(projectId, userId);

        if (projectOpt.isEmpty()) {
            return false;
        }

        projectRepository.delete(projectOpt.get());
        return true;
    }

    /**
     * Get public projects for discovery
     */
    @Transactional(readOnly = true)
    public List<ProjectDTO> getPublicProjects(int limit) {
        List<Project> projects = projectRepository.findByIsPublicTrueOrderByCreatedAtDesc();
        return projects.stream()
                .limit(limit)
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    /**
     * Convert Project entity to DTO with statistics
     */
    private ProjectDTO convertToDTO(Project project) {
        // Calculate statistics
        int totalQuestions = project.getQuestionSets().stream()
                .mapToInt(qs -> qs.getQuestions().size())
                .sum();

        // Get unique answered questions across all sessions
        Set<Long> answeredQuestionIds = new HashSet<>();
        LocalDateTime lastStudied = null;

        for (var questionSet : project.getQuestionSets()) {
            for (StudySession session : questionSet.getStudySessions()) {
                for (SessionAnswer answer : session.getSessionAnswers()) {
                    answeredQuestionIds.add(answer.getQuestionId());
                }

                // Find the most recent study session
                if (session.getCompletedAt() != null) {
                    if (lastStudied == null || session.getCompletedAt().isAfter(lastStudied)) {
                        lastStudied = session.getCompletedAt();
                    }
                }
            }
        }

        int questionsAnswered = answeredQuestionIds.size();
        double completionRatio = totalQuestions > 0 ? (double) questionsAnswered / totalQuestions : 0.0;

        return ProjectDTO.builder()
                .id(project.getId())
                .name(project.getName())
                .description(project.getDescription())
                .isPublic(project.getIsPublic())
                .ownerId(project.getOwnerId())
                .totalQuestions(totalQuestions)
                .questionsAnswered(questionsAnswered)
                .completionRatio(completionRatio)
                .lastStudied(lastStudied)
                .createdAt(project.getCreatedAt())
                .updatedAt(project.getUpdatedAt())
                .build();
    }
}
